packages/frontend/app/routes/payments._index.tsx

```diff
@@ -0,0 +1,177 @@
+  const searchParams = Object.fromEntries(url.searchParams.entries())
+
+  const result = paymentsSearchParams.safeParse(
+    searchParams.type
+      ? { ...searchParams, type: searchParams.type.split(',') }
+      : searchParams
+  )
+
+  if (!result.success) {
+    throw json(null, { status: 400, statusText: 'Invalid result.' })
+  }
+
+  const { type, ...pagination } = result.data
+
+  const payments = await listPayments({
+    ...pagination,
+    ...(type ? { filter: { type: { in: type } } } : {})
+  })
+
+  let previousPageUrl = '',
+    nextPageUrl = ''
+
+  if (payments.pageInfo.hasPreviousPage) {
+    previousPageUrl = `/payments?before=${payments.pageInfo.startCursor}`
+  }
+
+  if (payments.pageInfo.hasNextPage) {
+    nextPageUrl = `/payments?after=${payments.pageInfo.endCursor}`
+  }
+
+  return json({ payments, previousPageUrl, nextPageUrl, type })
+}
+
+export default function PaymentsPage() {
+  const { payments, previousPageUrl, nextPageUrl, type } =
+    useLoaderData<typeof loader>()
+  const setSearchParams = useSearchParams()[1]
+
+  const navigate = useNavigate()
+
+  function setTypeFilterParams(selectedType: PaymentType): void {
+    const selected = type
+
+    if (!selected.includes(selectedType)) {
+      selected.push(selectedType)
+      setSearchParams(new URLSearchParams({ type: selected.join(',') }))
+      return
+    }
+
+    setSearchParams(() => {
+      const newParams = selected.filter((t) => t !== selectedType).join(',')
+      if (newParams.length === 0) {
+        return ''
+      }
+
+      return new URLSearchParams({
+        type: newParams
+      })
+    })
+  }
+
+  const paymentSubpathByType: {
+    [key in PaymentType]: string
+  } = {
+    [PaymentType.Incoming]: 'incoming',
+    [PaymentType.Outgoing]: 'outgoing'
+  }
+
+  return (
+    <div className='pt-4 flex flex-col space-y-8'>
+      <div className='flex flex-col rounded-md bg-offwhite px-6'>
+        <PageHeader>
+          <div className='flex-1'>
+            <h3 className='text-2xl'>Payments</h3>
+          </div>
+        </PageHeader>
+        <div className='p-1'>
+          <h3 className='text-lg font-bold'>Filters</h3>
+          <div className='flex items-center'>
+            <PopoverFilter
+              label='Payment type'
+              values={type.length > 0 ? type : ['all']}
+              options={[
+                {
+                  name: 'All',
+                  value: 'all',
+                  action: () => {
+                    navigate(``)
+                  }
+                },
+                ...Object.values(PaymentType).map((value) => ({
+                  name: capitalize(value),
+                  value: value,
+                  action: () => {
+                    setTypeFilterParams(value)
+                  }
+                }))
+              ]}
+            />
+          </div>
+        </div>
+        <Table>
+          <Table.Head columns={['ID', 'Type', 'State']} />
```

Comment:
should we add `Date` here, to make it easier to see which one is most recent?

Reply:
yes, added